File: spaceship_crm/java8/spaceship_crm_java8_rest_jooq/src/main/java/com/spaceship/crm/controller/UserActivityRestController.java
/**
 * Author veni.vidi.dev (veni.vidi.dev@gmail.com)
 */
package com.spaceship.crm.controller;

import com.spaceship.crm.bean.LoginResponseBean;
import com.spaceship.crm.bean.UserActivityBean;
import com.spaceship.crm.entity.Activity;
import com.spaceship.crm.entity.User;
import com.spaceship.crm.service.ActivityService;
import com.spaceship.crm.service.SecurityService;
import com.spaceship.crm.service.UserActivityService;
import com.spaceship.crm.service.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping(path = "/rest")
@CrossOrigin
public class UserActivityRestController {

    @Autowired
    private SecurityService securityService;

    @Autowired
    private UserActivityService userActivityService;

    @Autowired
    private UserService userService;

    @Autowired
    private ActivityService activityService;

    @PostMapping("/login")
    public ResponseEntity<LoginResponseBean> doLogin(@RequestBody User user) {
        Optional<LoginResponseBean> loggedInUser = securityService.doLogin(user);
        if (loggedInUser.isPresent()) {
            return ResponseEntity.ok(loggedInUser.get());
        }
        return ResponseEntity.badRequest().build();
    }

    @GetMapping(path = "/activities", produces = MediaType.APPLICATION_JSON_VALUE)
    public List<UserActivityBean> getUserActivities() {
        return userActivityService.getAll();
    }

    @PostMapping(path = "/activity/{activityId}/user/{userId}", produces = MediaType.APPLICATION_JSON_VALUE)
    public List<UserActivityBean> addUserActivity(@PathVariable Long activityId, @PathVariable Long userId) {
        return userActivityService.addUserToActivityAndReturnAll(activityId, userId);
    }

    @GetMapping(path = "/allactivities", produces = MediaType.APPLICATION_JSON_VALUE)
    public List<Activity> getAllActivities() {
        return activityService.getAll();
    }

    @GetMapping(path = "/allusers", produces = MediaType.APPLICATION_JSON_VALUE)
    public List<User> getAllUsers() {
        return userService.getAll();
    }
    
}

File: spaceship_crm/scala/spaceship_crm_scala_spring_rest_jpa/src/main/scala/com/spaceship/crm/controller/UserActivityRestController.scala
/**
 * Author veni.vidi.dev (veni.vidi.dev@gmail.com)
 */
package com.spaceship.crm.controller

import java.util

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.{MediaType, ResponseEntity}
import org.springframework.web.bind.annotation._
import com.spaceship.crm.bean.{LoginResponseBean, UserActivityBean}
import com.spaceship.crm.entity.User
import com.spaceship.crm.security.SecurityService
import com.spaceship.crm.service.UserActivityService

@RestController
@RequestMapping(path = Array("/rest"))
@CrossOrigin
class UserActivityRestController(
                                  @Autowired private val securityService: SecurityService,
                                  @Autowired private val userActivityService: UserActivityService
                                ) {

  @PostMapping(path = Array("/login"))
  def doLogin(@RequestBody user: User): ResponseEntity[LoginResponseBean] = {
    val loggedInUser: LoginResponseBean = securityService.doLogin(user)
    ResponseEntity.ok(loggedInUser)
  }

  @GetMapping(
    path = Array("/activities"),
    produces = Array(MediaType.APPLICATION_JSON_VALUE)
  )
  def getUserActivities: util.List[UserActivityBean] = userActivityService.getAll

  @PostMapping(
    path = Array("/activity/{activityId}/user/{userId}"),
    produces = Array(MediaType.APPLICATION_JSON_VALUE)
  )
  def addUserActivity(@PathVariable activityId: Long, @PathVariable userId: Long): util.List[UserActivityBean] = {
    userActivityService.addUserToActivityAndReturnAll(activityId, userId)
  }
}
